File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
public:
    vector<int> arr;
    stack<int> minIdx;

    MinStack() {
        
    }
    
    void push(int val) {
        arr.push_back(val);
        int idx = arr.size() - 1;

        if (minIdx.size() == 0) {
            minIdx.push(idx);
        } else {
            if (val < arr[minIdx.top()]) {
                minIdx.push(idx);
            }
        }   
    }
    
    void pop() {
        int idx = arr.size() - 1;
        arr.pop_back();

        if (minIdx.top() == idx) {
            minIdx.pop();
        }
    }
    
    int top() {
        return arr[(int)arr.size() - 1];
    }
    
    int getMin() {
        return arr[minIdx.top()];
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack/0155-min-stack.java
class MinStack {

    Stack<Pair> st;
    public MinStack() {
        st = new Stack();
    }
    
    public void push(int val) {
        if (st.size() == 0) {
            Pair p = new Pair(val, val);
            st.push(p);
        } else {
            Pair top = st.peek();
            int mn = Math.min(top.minTillNow, val);
            Pair p = new Pair(val, mn);
            st.push(p);
        }
    }
    
    public void pop() {
        st.pop();
    }
    
    public int top() {
        Pair top = st.peek();
        return top.val;
    }
    
    public int getMin() {
        Pair top = st.peek();
        return top.minTillNow;
    }

    class Pair {
        int val;
        int minTillNow;

        public Pair(int v, int m) {
            this.val = val;
            this.minTillNow = m;
        }
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */